chapters/classes.tex

```diff
@@ -276,36 +276,127 @@ \subsubsection{Prefix Rules}\label{prefix-rules}
+\item
+  A \emph{continuous-time} variable \lstinline!vn! may have a non-vanishing time
+  derivative (\lstinline!der(vn)<>0! possible) and may also
+  change its value discontinuously at any time during transient analysis
+  (see \cref{events-and-synchronization}). If there are any discontinuities the variable is
+  not differentiable.
+\end{itemize}
+
+\begin{nonnormative}
+With every normal parameter, there is at least one reason why it isn't a structural parameter.
+This information is useful to maintain in tools, as it allows generation of informative error messages when a violation of structural expression variability is detected.
+For example:
+\begin{lstlisting}[language=modelica]
+parameter Integer n =
+  if b then 1 else 2;        /* Normal parameter due to cycle with 'b'. */
+parameter Boolean b = n > 3; /* Normal parameter due to cycle with 'n'. */
+Real[n] x;                   /* Variability error: 'n' must be structural. */
+\end{lstlisting}
+Here, a good error message for the variability error can include the information that the reason for \lstinline!n! being a normal parameter is that it is in a cycle with \lstinline!b!.
+
+(Would structural and normal parameters have had different variability prefixes in their component declarations, it would have been possible to detect and report errors directly at the structural component declarations instead.)
+\end{nonnormative}
+
+\begin{nonnormative}
+For a parameter in a valid model, presence of \lstinline!Evaluate! (\cref{modelica:Evaluate}) makes it possible to tell immediately whether it is a structural or normal parameter, at least as long as the warning described in \cref{modelica:Evaluate} isn't triggered.
+To see this, note that \lstinline!Evaluate = false! makes it a normal parameter by definition, and that \lstinline!Evaluate = true! would trigger the warning if the parameter is normal.
+\end{nonnormative}
+
+By the acyclic binding rule in \cref{acyclic-bindings-of-constants-and-parameters}, it follows that the value of a constant or structural parameter to be used in simplifications is possible to obtain by evaluation of a structural expression where values are available for all component subexpressions.
+
+\begin{example}
+A particularly demanding aspect of this evaluation is the potential presence of external functions.
+Hence, if it is known that a parameter won't be used by a structural expression, a user can make it clear that the external function is not meant to be evaluated during translation by using \lstinline!Evaluate = false!:
+\begin{lstlisting}[language=modelica]
+import length = Modelica.Utilities.Strings.length; /* Pure external function */
+parameter Integer n = length("Hello");             /* Structural parameter */
+parameter Integer p = length("Hello")
+  annotation(Evaluate = false);                    /* Normal parameter */
+parameter Boolean b = false;
+
+/* Variability analysis might cause evaluation of n: */
```

Comment:
What is *variability analysis*?

Reply:
The process of determining the variability of components and expressions.  This is needed to detect variability violations, and is now proposed to also be part of the classification of parameters as of structural or normal.

Reply:
It is a good name. Is it defined already?

Reply:
No, and now it isn't present in the example anymore.  Resolving.